### PR TITLE
Preliminary pagination support API

### LINKED ISSUE #59: Project listing agains taiga.io fails: too many projects
Hello,

I noticed that if you use python-taiga against https://api.taiga.io/ to list projects it fails.

```python
api.projects.list()
(Traceback (most recent call last):
  File "./taiga-stats", line 175, in cmd_list_projects
    for proj in api.projects.list():
  File "/home/erikw/.virtualenvs/taiga-stats/lib/python3.6/site-packages/taiga/models/base.py", line 41, in list
    objects = self.parse_list(result.json())
  File "/home/erikw/.virtualenvs/taiga-stats/lib/python3.6/site-packages/requests/models.py", line 892, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

No output fed in to the JSON decoding? So I decided to get down to my friend command line, and this works as expected:

```bash
$ curl -X GET -H "Content-Type: application/json" -H "Authorization: Bearer ${AUTH_TOKEN}" -s https://api.taiga.io/api/v1/projects
...
<project JSON>
```

So I drilled down the code and noticed that python-taiga [always send the header](https://github.com/nephila/python-taiga/blob/a4afbc7d795882d559bb32d450160fa05444eaca/taiga/requestmaker.py#L85)

```http
x-disable-pagination: True
```
to disable pagination and get everything at once.

So I plugged this in to the the previous command line

```bash
$ curl -X GET -H "Content-Type: application/json" -H "Authorization: Bearer ${AUTH_TOKEN}" -H "x-disable-pagination: True" -s https://api.taiga.io/api/v1/projects
<html>
<head><title>504 Gateway Time-out</title></head>
<body bgcolor="white">
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx/1.10.3 (Ubuntu)</center>
</body>
</html>
```

and now it it also fails like when called with pyhon-taiga.

I simply think this is the situation: now the number of public projects at taiga.io has recently grown too large to be included in one GET request to the project listing API endpoint. But before it was small enough to work.

A user of my project [taiga-stats](https://github.com/erikw/taiga-stats) reported this last week.

I guess python-taiga should be adapted to use pagination for requests, so it can handle large taiga instances, like taiga.io now is. And I think this is prioritized to to, as I guess taiga.io is the most used instance of Taiga of them all!

### PR DESCRIPTION
**Pull request by yakky**
This implements support for pagination
Fix #59 

By default `list` method retrieves all the objects navigating the pages returned by the API
`all` argument allows to disable pagination and return all the results (implementing current behavior)
`lazy_pagination` argument allows to load a single page

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -169,6 +169,27 @@ You can also specify filters
 
     tasks = api.tasks.list(project=1)
 
+By defaults list returns all objects, eventually getting the
```
Comment (erikw) [inline docs/index.rst]: s/By defaults/By default,/

File: docs/index.rst [hunk 2]
```diff
@@ -169,6 +169,27 @@ You can also specify filters
...
+
+.. warning:: be aware that if the unpaginated results may exceed
+             the data the parser can handle and may result in an error.
+
+You can also explicitly control the pagination by passing `lazy_pagination`
+argument and the page number
+
+.. code:: python
+
+   tasks_page_1 = api.tasks.list(lazy_pagination)  # Will only return page 1
```
Comment (erikw) [inline docs/index.rst]: Add example to get everything

**Comment by yakky** [comment]
@erikw thanks for your feedback. Does it look a sane API to you?

File: docs/index.rst (hunk 2 above)
Comment (yakky) [reply docs/index.rst]: a plain `api.tasks.list()` will still return all the results, like it does now, but using the pagination to get all the results

File: docs/index.rst (hunk 1 above)
Comment (yakky) [reply docs/index.rst]: Thanks, fixed

**Comment by yakky** [comment]
@erikw I added the `page_size` argument to allow improved control over the page size (and hence performance)

**Comment by erikw** [comment]
@yakky I was thinking it was a bit confusing with the different combinations you can make with `all` and `lazy_pagination`, but it's needed if we want to support all use cases I guess.

Maybe a matrix in the documentation like this one could help? 

| all        | lazy_pagination           | Output  |
| ------------- |:-------------:| -----:|
| False      | False | ...|
| False      | True | ...|
| True      | False | ...|
| True      | True | ...|

or possible reduce the number of use cases to support to maybe only have one input variable (with page number var too)  `paginated=True|False`?

**Comment by erikw** [comment]
@yakky Looks good! I was thinking it was a bit confusing though with the different combinations you can make with `all` and `lazy_pagination`, but it's needed if we want to support all use cases I guess.

Maybe a matrix in the documentation like this one could help? 

| all        | lazy_pagination           | Output  |
| ------------- |:-------------:| -----:|
| False      | False | ...|
| False      | True | ...|
| True      | False | ...|
| True      | True | ...|

or possible reduce the number of use cases to support to maybe only have one input variable (with page number var too)  `paginated=True|False`?

**Comment by yakky** [comment]
thanks for your feedback. We can use a 3-state argument (let's call it `pagination`), in addition to `page` argument to retrieve a specific page but I am worried it might be confusing:

| `pagination` | `page`       |   Output  |
| ------------- |------------| ----|
|  `True` (default)    |   `None`      |  All results retrieved by using paginated results and loading them behind the scenes |
|  `True` (default)    |   `<integer>`  |  Only results for the given page are retrieved |
| `False`                  | `unused`       |   Current behavior: all results, ignoring pagination |
| `<integer>`            |   `None`      |  All results retrieved by using paginated results and loading them behind the scenes, using given page size (higher page size could yield better performances) |
|  `<integer>`    |   `<integer>`  |  Only results for the given page of the given size are retrieved |

**Comment by erikw** [comment]
@yakky I agree, mixing types even for the same variable is not good. As long as it's clear or alternatively documented what will happen for which values of the argument, it's all good :)

**Comment by yakky** [comment]
@erikw how about this signature?

| `pagination` | `page_size` | `page`       |   Output  |
| --------------- |-----------------| --------------|------------|
|  `True` (default)    |   `<integer>` (default: 100; non numeric values fallbacks to the default)       |   `None`      |  All results retrieved by using paginated results and loading them behind the scenes, using given page size (higher page size could yield better performances) |
|  `True` (default)    |   `<integer>` (default: 100)       |   `<integer>`      |  Only results for the given page of the given size are retrieved |
| `False`                  | `unused`        | `unused`       |   Current behavior: all results, ignoring pagination |

**Comment by yakky** [comment]
@fmarco could you also give a look at this plz?

**Comment by erikw** [comment]
@yakky this last one looks sounder to me!

**Comment by yakky** [comment]
Thanks @erikw we will complete this with some tests in the next days and we'll likely release a new version sometimes next week

File: docs/index.rst [hunk 3]
```diff
@@ -169,6 +169,55 @@ You can also specify filters
...
+Examples
+^^^^^^^^^
+
+**No pagination**
+
+.. code:: python
+
+   tasks = api.tasks.list(paginate=False)
+
+.. warning:: be aware that if the unpaginated results may exceed
```
Comment (carloratm) [inline docs/index.rst]: Remove `if`?

File: taiga/models/base.py [hunk 4]
```diff
@@ -34,11 +34,55 @@ def __init__(self, requester):
...
+        """
+        Retrieves a list of objects.
+
+        By default uses local cache and remote pagination
+
+        If pagination is used and no page is requested (the default), all the
+        remote objects are retrieved and appended in a single list.
+
+        If pagination is disabled, all the objects are fetched from the
+        endpojnt and returned. This may trigger some parsing error if the
```
Comment (carloratm) [inline taiga/models/base.py]: `endpoint`

**Comment by yakky** [comment]
Thanks y'all. I'll fix those and I'll implement the tests for this. Hopefully we will be able to merge this and release a new version

**Comment by yakky** [comment]
@carloratm @fmarco could you have a look at the tests?

**Comment by erikw** [comment]
Awesome!

When can we expect a new release of python-taiga that includes this?

**Comment by yakky** [comment]
@erikw with a bit of luck within the weekend :crossed_fingers:

**Comment by erikw** [comment]
@yakky or maybe this weekend? :) 🙏

**Comment by yakky** [comment]
@erikw early next week 🙇  (but 0.9rc1 is already on PyPi)
0.9 will likely be identical

**Comment by yakky** [comment]
@erikw 0.9.0 released!

**Comment by erikw** [comment]
@yakky Many thanks!

I've now updated my project https://github.com/erikw/taiga-stats to use this release :)

**Comment by yakky** [comment]
Nice tool!
Do you mind adding a comment here #67 to list it?
I would like to start a section of tools and libraries using python-taiga to help people discovery nice stuff

**Comment by erikw** [comment]
Good idea!

**Pull request merged**